### PR TITLE
Reproducible Benchmarks

### LINKED ISSUE #13: Use the correct units in benchmarks

### PR DESCRIPTION
**Pull request by avik-pal**
Fixes #13 

TODOs:
- [x] Use deterministic point cloud generators
- [x] Use non-inplace versions of operators
- [x] Update Kaolin scripts

### CODE REVIEW

**Comment by DhairyaLGandhi** [comment]
bump

**Comment by nirmal-suthar** [comment]
I am also changing the format of PointCloud to 3xNxB, so this benchmark need to be updated. I will make necessary changes to the benchmark once I finish this change.

**Comment by nirmal-suthar** [comment]
@avik-pal  I have added kaolin equivalent script for benchmark, can you please review it.

**Comment by avik-pal** [comment]
LGTM. Can you run the benchmarks and report the results here before it goes in?

**Comment by nirmal-suthar** [comment]
(Kaolin)
```
DEVICE: CPU
Running benchmarks for npoints = 4096
ScalePointCloud: 0.061511993408203125 ms
RotatePointCloud: 0.049591064453125 ms
RealignPointCloud: 8.382320404052734 ms
NormalizePointCloud: 0.48279762268066406 ms
Compose: 0.6380081176757812 ms

Running benchmarks for npoints = 16384
ScalePointCloud: 7.851362228393555 ms
RotatePointCloud: 7.779836654663086 ms
RealignPointCloud: 34.340858459472656 ms
NormalizePointCloud: 23.39315414428711 ms
Compose: 70.2211856842041 ms

Running benchmarks for npoints = 65536
ScalePointCloud: 7.91621208190918 ms
RotatePointCloud: 7.879734039306641 ms
RealignPointCloud: 34.55471992492676 ms
NormalizePointCloud: 23.876667022705078 ms
Compose: 70.770263671875 ms

Running benchmarks for npoints = 262144
ScalePointCloud: 8.008003234863281 ms
RotatePointCloud: 8.074760437011719 ms
RealignPointCloud: 39.08681869506836 ms
NormalizePointCloud: 26.52740478515625 ms
Compose: 75.27351379394531 ms

Running benchmarks for npoints = 1048576
ScalePointCloud: 9.147882461547852 ms
RotatePointCloud: 9.363174438476562 ms
RealignPointCloud: 74.33509826660156 ms
NormalizePointCloud: 48.63238334655762 ms
Compose: 108.71624946594238 ms

CUDA is on. Running GPU Benchmarks
DEVICE: GPU
Running benchmarks for npoints = 4096
ScalePointCloud: 0.19551999866962433 ms
RotatePointCloud: 0.10787200182676315 ms
RealignPointCloud: 0.9790719747543334 ms
NormalizePointCloud: 0.23318399488925934 ms
Compose: 1.2248640060424805 ms

Running benchmarks for npoints = 16384
ScalePointCloud: 0.1963520050048828 ms
RotatePointCloud: 0.1074879989027977 ms
RealignPointCloud: 3.4454400539398193 ms
NormalizePointCloud: 0.2341119945049286 ms
Compose: 3.7479360103607178 ms

Running benchmarks for npoints = 65536
ScalePointCloud: 0.19500799477100372 ms
RotatePointCloud: 0.12303999811410905 ms
RealignPointCloud: 13.480992317199707 ms
NormalizePointCloud: 0.25571200251579285 ms
Compose: 13.843168258666992 ms

Running benchmarks for npoints = 262144
ScalePointCloud: 0.19657599925994873 ms
RotatePointCloud: 0.30796799063682556 ms
RealignPointCloud: 80.64806365966797 ms
NormalizePointCloud: 0.25577598810195923 ms
Compose: 81.30508422851562 ms

Running benchmarks for npoints = 1048576
ScalePointCloud: 0.2795200049877167 ms
RotatePointCloud: 1.0577919483184814 ms
RealignPointCloud: 324.3732604980469 ms
NormalizePointCloud: 0.3883199989795685 ms
Compose: 326.091796875 ms
```

**Comment by DhairyaLGandhi** [comment]
How are you running the benchmarks?

**Comment by nirmal-suthar** [comment]
> How are you running the benchmarks?

on cyclops

**Comment by nirmal-suthar** [comment]
```

**Comment by DhairyaLGandhi** [comment]
Cyclops can be a bit noisy, but more than that, use `timeit` to get your times.

**Comment by DhairyaLGandhi** [comment]
Also please add graphs for these numbers to the readme.

File: src/transforms/pcloud_func.jl [hunk 1]
```diff
@@ -107,7 +107,7 @@ function rotate!(pcloud::PointCloud, rotmat::AbstractArray{Float32,2})
     size(rotmat) == (3, 3) || error("rotmat must be (3, 3) array, but instead got $(size(rotmat)) array")
     size(pcloud.points, 1) == 3 || error("dimension of points in PointCloud must be 3")
     points_packed = transpose(rotmat) * reshape(pcloud.points, 3, :)
-    pcloud.points = reshape(points_packed, size(pcloud.points)...)
+    pcloud.points = copy(reshape(points_packed, size(pcloud.points)...))
```
Comment (nirmal-suthar) [inline src/transforms/pcloud_func.jl:110]: @DhairyaLGandhi @avik-pal when using CuArray, `reshape` changes parent field of CuArray, which causes error in deepcopy. Here using `copy` degrades the performance.

File: src/transforms/pcloud_func.jl [hunk 2]
```diff
@@ -57,7 +57,7 @@ julia> scale!(p, 1.0)
 """
 function scale!(pcloud::PointCloud, factor::Float32)
     (factor > 0.0) || error("factor must be greater than 0.0")
-    pcloud.points = pcloud.points .* factor
+    lmul!(factor, pcloud.points)
```
Comment (nirmal-suthar) [inline src/transforms/pcloud_func.jl:60]: for some reason lmul! is twice as better than using vectorized operation

File: src/transforms/pcloud_func.jl (hunk 1 above)
Comment (DhairyaLGandhi) [reply src/transforms/pcloud_func.jl:110]: Are you thinking of a `ReshapedArray` here? That's when the parent would be transformed.

Comment (nirmal-suthar) [reply src/transforms/pcloud_func.jl:110]: Currently Yes. The exact error is, after reshaping, pcloud.points is `CuArray{Float32, 2, CuArray{Float32, 2, nothing}}` and when applying deepcopy on this `pcloud`, new points changes to `CuArray{Float32, 2, nothing}` and ultimately fails at deepcopy_internal as both types don't match exactly.

**Comment by nirmal-suthar** [comment]
Final graphs, I think using horizontal would be better. 

bm_plot1
bm_plot2

**Comment by avik-pal** [comment]
The horizontal plot is definitely the one we should go with

**Comment by nirmal-suthar** [comment]
I am thinking to add TriMesh benchmarks and graph in this PR before merging. Will rebase this once TriMesh PR is merged.

**Comment by avik-pal** [comment]
Maybe let's merge this PR before the TriMesh. We might want to do performance regressions in the benchmark PR for Meshes.

Also, don't forget to add the plots in the README

**Comment by nirmal-suthar** [comment]
Sure

**Comment by avik-pal** [comment]
@nirmal-suthar You missed adding the plot to README

**Comment by nirmal-suthar** [comment]
I have updated readme on master, missed it here.

**Comment by DhairyaLGandhi** [comment]
Pretty sure it's on the readme rn

**Pull request merged**